MudDataGrid: Add CancellationToken support for ServerData

w3ori opened the pull request:
## Description
Add CancellationToken support for ServerData with to prevent unnecessary data loading.
My main issue was with rapidly go over multiple pages and it loaded the data for every skipped page.

Try to resolve: #11185

## Type of Changes
- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [x]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ]  Documentation (fix or improvement to the website or code docs)

## Checklist
- [x]  The PR is submitted to the correct branch (dev).
- [x]  My code follows the code style of this project.
- [ ]  I've added relevant tests.
- [x]  I've fixed/modified existing relevant tests.

w3ori committed:
Enhance MudDataGrid ServerData with CancellationToken support

w3ori committed:
Fix MudDataGrid examples to use CancellationToken for ServerData

w3ori committed:
Add CancellationToken to Task.Delay() in DataGridServerSideExample

w3ori commented on src/MudBlazor/Components/DataGrid/MudDataGrid.razor.cs:
```diff
@@ -957,7 +957,7 @@ private void ApplyInitialExpansionForItems(IEnumerable<T> items)
         /// </remarks>
         [Parameter]
-        public Func<GridState<T>, Task<GridData<T>>> ServerData { get; set; }
+        public Func<GridState<T>, CancellationToken, Task<GridData<T>>> ServerData { get; set; }
```
XML documentation is same detailed as for `VirtualizeServerData`.

w3ori commented on src/MudBlazor.Docs/Pages/Components/DataGrid/Examples/DataGridServerSideExample.razor:
```diff
@@ -38,9 +38,9 @@
     {
-        IEnumerable<Element> data = await httpClient.GetFromJsonAsync<List<Element>>("webapi/periodictable");
+        IEnumerable<Element> data = await httpClient.GetFromJsonAsync<List<Element>>("webapi/periodictable", token);
         await Task.Delay(300);
```
Fixed.

w3ori committed:
Fix tests

w3ori committed:
Fix more tests

w3ori committed:
Set culture for DataGridGroupingTemplateSetAtGridLevel test method

Adds `[SetCulture("en-US")]` attribute to ensure consistent number formatting during test execution.

w3ori commented:
/gemini review

w3ori commented on src/MudBlazor/Components/DataGrid/MudDataGrid.razor.cs:
```diff
@@ -1472,7 +1472,10 @@ internal async Task InvokeServerLoadFunc()
 
-                _serverData = await ServerData(state);
+                // Cancel any prior request
+                CancelServerDataToken();
```
I didn't modify the original `CancelServerDataToken()` method.

w3ori commented on src/MudBlazor.Docs/Pages/Components/DataGrid/Examples/DataGridGroupingMultiLevelExample.razor:
```diff
@@ -196,7 +196,7 @@
     }
 
-    private async Task<GridData<USState>> ServerReload(GridState<USState> state)
+    private async Task<GridData<USState>> ServerReload(GridState<USState> state, CancellationToken token)
```
The name was the same before I made any changes. I can fix it if the maintainer requests it.

w3ori commented on src/MudBlazor.Docs/Pages/Components/DataGrid/Examples/DataGridServerSideExample.razor:
```diff
@@ -38,10 +38,10 @@
     /// Here we simulate getting the paged, filtered and ordered data from the server
     /// </summary>
-    private async Task<GridData<Element>> ServerReload(GridState<Element> state)
+    private async Task<GridData<Element>> ServerReload(GridState<Element> state, CancellationToken token)
```
The name was the same before I made any changes. I can fix it if the maintainer requests it.

w3ori commented on src/MudBlazor.UnitTests.Viewer/TestComponents/DataGrid/DataGridFilterableServerDataTest.razor:
```diff
@@ -15,7 +15,7 @@
     };
 
-    private async Task<GridData<Item>> ServerDataFunc(GridState<Item> gridState)
+    private async Task<GridData<Item>> ServerDataFunc(GridState<Item> gridState, CancellationToken token)
```
The name was the same before I made any changes. I can fix it if the maintainer requests it.

w3ori commented on src/MudBlazor.UnitTests.Viewer/TestComponents/DataGrid/DataGridGroupExpandedFalseServerDataTest.razor:
```diff
@@ -31,9 +31,9 @@
     ];
 
-    private async Task<GridData<Fruit>> ServerDataFunc(GridState<Fruit> gridState)
+    private async Task<GridData<Fruit>> ServerDataFunc(GridState<Fruit> gridState, CancellationToken token)
```
The name was the same before I made any changes. I can fix it if the maintainer requests it.

w3ori commented on src/MudBlazor.UnitTests.Viewer/TestComponents/DataGrid/DataGridGroupExpandedServerDataTest.razor:
```diff
@@ -31,9 +31,9 @@
     ];
 
-    private async Task<GridData<Fruit>> ServerDataFunc(GridState<Fruit> gridState)
+    private async Task<GridData<Fruit>> ServerDataFunc(GridState<Fruit> gridState, CancellationToken token)
```
The name was the same before I made any changes. I can fix it if the maintainer requests it.

w3ori commented on src/MudBlazor.UnitTests.Viewer/TestComponents/DataGrid/DataGridGroupableServerDataTest.razor:
```diff
@@ -36,7 +36,7 @@
     };
 
-    private async Task<GridData<Item>> ServerDataFunc(GridState<Item> gridState)
+    private async Task<GridData<Item>> ServerDataFunc(GridState<Item> gridState, CancellationToken token)
```
The name was the same before I made any changes. I can fix it if the maintainer requests it.

w3ori commented on src/MudBlazor.UnitTests.Viewer/TestComponents/DataGrid/DataGridGroupingMultiLevelTest.razor:
```diff
@@ -194,12 +194,12 @@
     }
 
-    private async Task<GridData<USState>> ServerReload(GridState<USState> state)
+    private async Task<GridData<USState>> ServerReload(GridState<USState> state, CancellationToken token)
```
The name was the same before I made any changes. I can fix it if the maintainer requests it.

w3ori commented on src/MudBlazor.UnitTests.Viewer/TestComponents/DataGrid/DataGridHierarchyInitiallyExpandedServerDataTest.razor:
```diff
@@ -50,9 +50,9 @@
         };
 
-    private async Task<GridData<Model>> ServerReload(GridState<Model> state)
+    private async Task<GridData<Model>> ServerReload(GridState<Model> state, CancellationToken token)
```
The name was the same before I made any changes. I can fix it if the maintainer requests it.

w3ori commented on src/MudBlazor.UnitTests.Viewer/TestComponents/DataGrid/DataGridServerDataColumnFilterMenuTest.razor:
```diff
@@ -24,7 +24,7 @@
     };
 
-    private async Task<GridData<Model>> ServerDataFunc(GridState<Model> gridState)
+    private async Task<GridData<Model>> ServerDataFunc(GridState<Model> gridState, CancellationToken token)
```
The name was the same before I made any changes. I can fix it if the maintainer requests it.

w3ori commented on src/MudBlazor.UnitTests.Viewer/TestComponents/DataGrid/DataGridServerDataColumnFilterRowTest.razor:
```diff
@@ -23,7 +23,7 @@
     };
 
-    private async Task<GridData<Model>> ServerDataFunc(GridState<Model> gridState)
+    private async Task<GridData<Model>> ServerDataFunc(GridState<Model> gridState, CancellationToken token)
```
The name was the same before I made any changes. I can fix it if the maintainer requests it.

w3ori commented on src/MudBlazor.UnitTests.Viewer/TestComponents/DataGrid/DataGridServerMultiSelectionTest.razor:
```diff
@@ -21,7 +21,7 @@
     /// Here we simulate getting the data from the server
     /// </summary>
-    private static async Task<GridData<Item>> ServerReload(GridState<Item> state)
+    private static async Task<GridData<Item>> ServerReload(GridState<Item> state, CancellationToken token)
```
The name was the same before I made any changes. I can fix it if the maintainer requests it.

w3ori commented on src/MudBlazor.UnitTests.Viewer/TestComponents/DataGrid/DataGridServerPaginationTest.razor:
```diff
@@ -27,7 +27,7 @@
     }
 
-    private async Task<GridData<Model>> ServerReload(GridState<Model> state)
+    private async Task<GridData<Model>> ServerReload(GridState<Model> state, CancellationToken token)
```
The name was the same before I made any changes. I can fix it if the maintainer requests it.

w3ori commented on src/MudBlazor.UnitTests.Viewer/TestComponents/DataGrid/DataGridServerSideSortableTest.razor:
```diff
@@ -24,7 +24,7 @@
     /// Here we simulate getting the paged, filtered and ordered data from the server
     /// </summary>
-    private async Task<GridData<Item>> ServerReload(GridState<Item> state)
+    private async Task<GridData<Item>> ServerReload(GridState<Item> state, CancellationToken token)
```
The name was the same before I made any changes. I can fix it if the maintainer requests it.

ScarletKuro commented:
> Breaking changes can be merged only when a new main version comes out, for example `v9`?

yeah

versile2 commented:
Just leave here for now and since it's a breaking change, go ahead and change the name to append Async and add the ?.Dispose to CancelServerData. This will likely need to be updated to the latest dev occasionally.

V

w3ori committed:
Merge branch 'MudBlazor:dev' into feature/datagrid-server-side-paging-cancellationtoken

w3ori committed:
Rename MudDataGrid 'ServerData' to 'ServerDataAsync'

w3ori committed:
Dispose CancelServerData in CancelServerDataToken()

w3ori committed:
Fix Exception messages should be in tests

w3ori committed:
Rename ServerData to ServerDataAsync for MudDataGrid test components

w3ori commented:
> Just leave here for now and since it's a breaking change, go ahead and change the name to append Async and add the ?.Dispose to CancelServerData. This will likely need to be updated to the latest dev occasionally.
> 
> V

I renamed `ServerData` to `ServerDataAsync`.

Is Dispose on the correct place? https://github.com/MudBlazor/MudBlazor/pull/11841/commits/b43c3992c2da04c180cdc7f9d22369ed4b9caff0 

Should I rename test razor files also from ServerData to ServerDataAsync in filenames?

I also updated to recent `dev` branch.

ScarletKuro commented:
> go ahead and change the name to append Async

> I renamed `ServerData` to `ServerDataAsync`.

I don't think that's a good idea. Yes, normal properties or methods should have an `Async` prefix, but we're talking about a component parameter here. I haven't seen anyone use this in practice for a `[Parameter]`. Moreover, this now deviates from `MudTable`:
https://github.com/MudBlazor/MudBlazor/blob/bc08c8862324829a2c1b1d7fcbd3ce936227d623/src/MudBlazor/Components/Table/MudTable.razor.cs#L658

And I don't think having inconsistency is a good thing.

ScarletKuro commented:
> @ScarletKuro Should I rename it back?

Let's wait @versile2 @danielchalmers says on my comment.

w3ori committed:
Revert "Rename MudDataGrid 'ServerData' to 'ServerDataAsync'"

This reverts commit 9833df4e31b25ada91495a73f845c88191da524e.

w3ori committed:
Revert "Fix Exception messages should be in tests"

This reverts commit 4e01d0821a888a3c9bb2c1a1556f06071bc08898.

w3ori committed:
Revert "Rename ServerData to ServerDataAsync for MudDataGrid test components"

This reverts commit c9104119fa3bfb7074739bd77cbfb1e5f3826526.

w3ori committed:
Merge branch 'MudBlazor:dev' into feature/datagrid-server-side-paging-cancellationtoken

danielchalmers commented:
> yup was my bad on the naming. Can I blame it on the bot?

Having something to blame is actually its best feature!

w3ori committed:
Merge branch 'dev' into feature/datagrid-server-side-paging-cancellationtoken

w3ori committed:
Merge branch 'MudBlazor:dev' into feature/datagrid-server-side-paging-cancellationtoken

w3ori committed:
Merge branch 'MudBlazor:dev' into feature/datagrid-server-side-paging-cancellationtoken

danielchalmers commented:
Thank you!
